### DISCUSSION TITLE
`reduce`  window and aggregate functions

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#10703] [Ideas]
The ability to compute arbitrary [folds/reductions/accumulations/scans/recursions](https://en.wikipedia.org/wiki/Fold_(higher-order_function)), such as the much requested exponentially weighted moving averages (#12591) or finite state machines (#3994), in O(n) using window syntax would be a killer feature, likely to win over more pandas or polars users (neither of which has a great solution for this).

SQL window syntax and duckdb lambda functions are a match made in heaven to open up a wide class of window functions that are trivially implementable in `O(n)` as a window function, but require `O(n^2)` time and space in an awkward workaround using `list_reduce`.

Similarly, complicated stateful queries such as "date of longest streak" requested on [r/duckdb](https://www.reddit.com/r/DuckDB/comments/1o5iews/window_functions_in_group_by_batches_or_some/) could be implemented trivially with O(1) space and O(n) runtime complexity with a `reduce` aggregate function. 

---

For example, consider a `sales` table listing daily dollar amounts of products sold by each salesperson:
```sql
CREATE OR REPLACE TABLE sales (
    agent TEXT,
    dollars_sold FLOAT,
    date DATE
);
INSERT INTO sales VALUES 
('A', 100.0, '2020-01-01'),
('B', 150.0, '2020-01-01'),
('A', 10.0, '2020-01-04'), 
('B', 100.0, '2020-01-02'),
('A', 100.0, '2020-01-05'),
('B', 10.0, '2020-01-06');
```

Assume agents put their sales into an interest-earning account, so that every day their wealth increases by 5% interest plus whatever sales they made on that day.

With a `reduce` window function (I'll use the  name `reduce` from here on because duckdb already has `list_reduce`), this could be written as 
```sql
SELECT 
*,
date - lag(date) OVER cum AS days_passed,
reduce(0, wealth -> wealth * pow(1.05,  days_passed) + dollars_sold) OVER cum AS wealth
FROM sales
WINDOW cum AS (PARTITION BY agent ORDER BY date)
```
```
┌─────────┬────────────┬───────────┐
│  agent  │    date    │  wealth   │
│ varchar │    date    │   float   │
├─────────┼────────────┼───────────┤
│ A       │ 2020-01-01 │     100.0 │
│ A       │ 2020-01-04 │  125.7625 │
│ A       │ 2020-01-05 │ 232.05063 │
│ B       │ 2020-01-01 │     150.0 │
│ B       │ 2020-01-02 │     257.5 │
│ B       │ 2020-01-06 │ 322.99286 │
└─────────┴────────────┴───────────┘
```

At the moment, achieving the same result is cumbersome and requires `O(n^2)` time and space using `list_reduce`:
```sql
SELECT 
*,
list_reduce(
    list(struct_pack(dollars_sold, date)) OVER (PARTITION BY agent ORDER BY date), 
    (state, next) -> struct_pack(
        dollars_sold:=state.dollars_sold * pow(1.05, next.date - state.date) + next.dollars_sold,
        date:=next.date
    )
).dollars_sold AS wealth
FROM sales
```

I believe the newly proposed syntax is 
- more easily written 
- more easily readable
- more easily modifiable, e.g., to use a variable interest rate that is joined in from some other table
- more easily optimizable by duckdb, since a window-reduce (at least in the case of un-framed windows) will always be computable in O(n) time unlike a general `list_reduce(arbitrary-column-expression)` (I guess one could hand-tune `list_reduce` for when it's applied to a non-framed window function and the reduction lambda has an empty closure, but I don't see how that'd be better than having a window function)

PS:
- Compared to `list_reduce`, there is an extra argument for the initial value for full generality, and unlike for `list_reduce` the lambda only has one argument and every available column (expression) can be used in place of the second argument
- Because it might matter these days: the syntax proposed here is roughly what ChatGPT dreamed up already exists, showing it might be intuitive to users (or it already exists in related dialects)
- The aggregate function is not quite as important as the window function, since the equivalent `list_reduce(list_agg)` is going to be O(n) runtime as well, but since it can't easily be optimized it's currently unnecessarily using O(n) space
- For future reference, another extremely cumbersome solution, via CTE:
```sql
WITH RECURSIVE cte(agent, date, wealth) AS (
    (
        SELECT 
        DISTINCT ON (agent)
        agent,
        date,
        dollars_sold AS wealth,
        FROM sales
        ORDER BY date ASC
    )
    UNION ALL
    (
        SELECT 
        cte.agent,
        next_sale.date,
        cte.wealth * pow(1.05, next_sale.date - cte.date) + next_sale.dollars_sold
        FROM 
        cte, 
        LATERAL 
        (SELECT DISTINCT ON (agent) * FROM sales WHERE sales.date > cte.date AND sales.agent = cte.agent ORDER BY sales.date ASC) next_sale
    )
) 
SELECT * FROM cte ORDER BY agent, date
```

**Comment by corwinjoy** [comment]
@hawkfish Per our discussion at the conference.  What do you think of this idea?

**Comment by hawkfish** [comment]
I think the problem here is that our aggregation model can be used directly for windowing, but it is [more complex](https://github.com/duckdb/duckdb/blob/main/src/core_functions/aggregate/README.md#aggregation-operations) to allow for parallelism, which pandas does not support. The example provided could be partially shoehorned into the aggregation model like so:

* `size` would be the size of the datatype plus an initialised flag
* `init` would copy the starting value
* `update` would apply the lambda
* `finalize` would just copy the value out of the state, or a default value (`NULL`) if the flag was not set.

The trouble is that `combine` is not defined by the lambda, so we would be unable to parallelise the computation.

**Reply by soerenwolfers** [reply]
Being unable to parallelize the per-partition computation of a general reduce is inherent to the problem and not a restriction of any specific implementation; as such it shouldn't be held against this proposal but rather for it: whatever shoehorning is necessary to make this work in duckdb doesn't lose out compared to the theoretical optimum (at least not for this reason). 

Btw, in practice, everytime I run into this kind of problem, I'd get enough parallelism from the `PARTITION BY`, e.g., I have many more agents than I have cores.

**Comment by hawkfish** [comment]
Even if we treated lambdas as a new kind of non-aggregate window function, I don't see how chunk 2 in the example could even get started until chunk 1 had been computed. All the other window functions perform "global" calculations either on the row position or the fully materialised partition data. That is unfortunately not the case here.

**Reply by soerenwolfers** [reply]
See previous comment: yes, this window function requires sequential computation, but (a) that's unavoidable in the kind of problems this functionality would solve (b) practical performance is often dominated by having to do this on multiple partitions, which *can* be parallelized over (c) existing workarounds to this kind of functionality don't even have the correct O(n) complexity (the list_agg&list_reduce solution doesn't; the recursive CTE might, but it's practically slow beyond hope).

**Comment by hawkfish** [comment]
Oh I see the value - but it requires a partial redesign of the operators and planning code, instead of being a simple drop in to the evaluation subsystem.

**Comment by hawkfish** [comment]
When I was at Tableau, this kind of iteration was added as part of the "table calculations" suite, but when we started thinking about pushing them down to the database, we realised it doesn't play well with the SQL analytic function model. Which is not insurmountable, but it is not trivial either. 

It is all related to another thing hiding in the original example, which is reusable window functions (`days_passed`). Right now we end up generating two window operators with identical partitioning and ordering, but it would be nice to be able to share the partitions and just compute new functions. If we had that capability, then we could just pull out the single-threaded functions into separate pre-partitioned operators and still get most of the benefit of parallelism for the other functions.

**Comment by corwinjoy** [comment]
Just for reference, what @hawkfish is talking about here are the internal requirements for an aggregate in the duckdb code.
https://github.com/duckdb/duckdb/tree/main/src/core_functions/aggregate

**Reply by corwinjoy** [reply]
@soerenwolfers I spent some time taking a look to see if I could implement this, but it was beyond me for a few reasons:
1. From what I can see, the aggregate functions take a fixed number of columns and I am not sure how to write a function that would take a varying number of arguments for the variables used by the lambda. (This would be for a version of the function that looked like `reduce(wealth, days_passed, dollars_sold)`
2. And/or this would need some tricky work to get the correct execution context for the lambda. The existing scoping rules are oriented around a list, then table variables. (https://duckdb.org/docs/sql/functions/lambda.html). Having a context to access row level variables would take some work or munging to rewrite the lambda function.

As an alternative, I took a crack to see what I could do via macros. I was able to make some progress but I couldn't find a way to generalize the window expression either via CTE or argument. It might also be possible to create the lambda expression on the fly to simplify this lambda expression but I am not sure if that is supported. Here is what I have so far:

```
CREATE OR REPLACE MACRO dynamic_reduce(
        lambda_function
    ) AS TABLE (
    FROM reduce_cte
    SELECT 
        *,
        -- Use the list_reduce function to apply an aggregate
        -- function of our choice
        list_reduce(
            list(reduce_cte) OVER (PARTITION BY agent ORDER BY date),
            lambda_function
        ) lambda_result
);

WITH reduce_cte AS (
    SELECT * from sales
)
SELECT agent, dollars_sold, date, lambda_result.dollars_sold as wealth
FROM dynamic_reduce(
    (state, row) -> struct_pack(
                      agent:=row.agent,
                      dollars_sold:=row.dollars_sold * pow(1.05, row.date - state.date) + row.dollars_sold,
                      date:=row.date
                    )
);
```

```
┌─────────┬──────────────┬────────────┬───────────┐
│  agent  │ dollars_sold │    date    │  wealth   │
│ varchar │    float     │    date    │   float   │
├─────────┼──────────────┼────────────┼───────────┤
│ B       │        150.0 │ 2020-01-01 │     150.0 │
│ B       │        100.0 │ 2020-01-02 │     205.0 │
│ B       │         10.0 │ 2020-01-06 │ 22.155062 │
│ A       │        100.0 │ 2020-01-01 │     100.0 │
│ A       │         10.0 │ 2020-01-04 │  21.57625 │
│ A       │        100.0 │ 2020-01-05 │     205.0 │
└─────────┴──────────────┴────────────┴───────────┘

```

**Comment by jprafael** [comment]
Can anyone tell me whats the status here?

**Reply by soerenwolfers** [reply]
As far as I know, nobody is working on this

**Comment by hawkfish** [comment]
My current plan is to add window functions to the catalogue for 1.6. This would in turn allow you to write extensions that contain novel window functions such as `reduce`. It may also be that the new PEG parser will make it easier to add appropriate syntax, but that would have to result in parameters that can be stored in the catalogue.

I've already made one or two small changes in this direction on `main`, but it will be a while before it is ready for prime time.

**Comment by hawkfish** [comment]
I have started working on the in earnest and the changes have started appearing on `main`. It will be a while before I have the APIs fully defined as there are a lot of callbacks that the functions need to make (frame requirements, expression sharing, state management, streaming support and so on) but it is moving forward.

**Comment by hawkfish** [comment]
For all interested parties, the new window catalog/extension work has landed on `main`. There is a simple example in `loadable_extension_demo.cpp` that implements both the blocking and streaming APIs, and of course the standard functions have all been ported to the new APIs. Can't wait to see what everyone does!

**Comment by corwinjoy** [comment]
Very Cool! Looking forward to trying it out!

**Comment by hawkfish** [comment]
Closing this because the feature now complete on `main`. There is documentation in `src/function/window/README.md` for anyone who wants to get their hands dirty.
